docs(tutorials): add "Browser dApp with an injected wallet" tutorial (#1109)
midnightntwrk/midnight-docs: docs/tutorials/browser-dapp/index.md

---
SPDX-License-Identifier: Apache-2.0
copyright: This file is part of midnight-docs. Copyright (C) Midnight Foundation. Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
title: Browser dApp with an injected wallet

> description: Deploy a Compact contract, call a circuit, and read contract state from a minimal browser dApp that delegates sync, balancing, and signing to an injected wallet (Lace or 1AM) — no headless wallet sync required.

Reviewer: The `description` field is 212 characters; the style guide requires 120-160. Also replaces the em-dash and uses `DApp` per brand guide.

```suggestion
description: Deploy a contract, call circuits, and read on-chain state from a browser DApp using an injected wallet via the DApp Connector API.
```

> sidebar_position: 60

Reviewer: Every tutorial sub-page includes `toc_max_heading_level: 2`. Verified across bboard, bship, leaderboard, and zk-loan.

```suggestion
sidebar_position: 60
toc_max_heading_level: 2
```

tags:
  - tutorials
  - dapps
  - compact
  - wallet
  - midnight-js

>   - midnight-network

Reviewer: Existing tutorials include a difficulty tag (`beginner`, `intermediate`, `advanced`). This one reads as intermediate given the SDK and wallet prerequisites.

```suggestion
  - midnight-network
  - intermediate
```

---

> # Browser dApp with an injected wallet

Reviewer: "DApp" is the only accepted spelling per the brand guide (capital D, capital A). This applies throughout the file (title, description, H1, body text). Fixing H1 here; consider a find-and-replace for the remaining instances.

```suggestion
# Browser DApp with an injected wallet
```

> This tutorial shows how to **deploy a Compact contract, call a circuit, and read contract state** from a small browser dApp that delegates the heavy work — chain sync, balancing, proving, and signing — to an **injected wallet** (Lace or 1AM) through the DApp connector API.

Reviewer: Replacing em-dashes with parentheses and fixing `dApp` to `DApp`.

```suggestion
This tutorial shows how to **deploy a Compact contract, call a circuit, and read contract state** from a small browser DApp that delegates the heavy work (chain sync, balancing, proving, and signing) to an **injected wallet** (Lace or 1AM) through the DApp Connector API.
```

> Many examples run a *headless* wallet that syncs the chain inside your own Node process. A fresh wallet must scan the full shielded and DUST history, which is memory-intensive and can fail before it completes on a typical machine. By letting the browser wallet sync and balance instead, your dApp only builds the transaction and asks the wallet to finalize it — so your app never runs that expensive sync. This is the same model the bulletin-board tutorial uses, generalized here to deploy, call, and read.

Reviewer: Replacing em-dash with comma. Also, the sentence starting "A fresh wallet must scan..." runs to 36 words. Consider splitting after "typical machine" for readability.

```suggestion
Many examples run a *headless* wallet that syncs the chain inside your own Node process. A fresh wallet must scan the full shielded and DUST history, which is memory-intensive and can fail before it completes on a typical machine. By letting the browser wallet sync and balance instead, your DApp only builds the transaction and asks the wallet to finalize it, so your app never runs that expensive sync. This is the same model the bulletin-board tutorial uses, generalized here to deploy, call, and read.
```

> This guide targets the current SDK line: `midnight-js` 4.x, `compact-runtime` 0.15, ledger v8. It works on **preview** and **preprod**.

Reviewer: Worth noting that `compact-runtime` 0.16.0 is now the latest published version (`npm view @midnight-ntwrk/compact-runtime@latest version` returns `0.16.0`). The dependency list on line 37 pins `^0.15.0`, which in semver resolves to `>=0.15.0 <0.16.0` and will not install 0.16.0. Consider updating both the prose and the dependency to 0.16 to prevent version drift on day one.

## Prerequisites

- **Node** v22.17.1 or higher, and a Vite app.
- A **compiled Compact contract**: a `managed/<name>/` directory containing `contract/index.js` (generated bindings) plus the ZK assets `keys/` and `zkir/`. Pin `compactc`, the language version, and the runtime together.
- A **wallet extension** (Lace or 1AM) installed in your browser, unlocked, set to **preview** (or preprod), with the connected account funded with tNIGHT and registered for DUST (DUST pays fees).

> - A **proof server** — either your own (`docker run -p 6300:6300 midnightntwrk/proof-server:<tag>`) or the one the wallet reports via `getConfiguration()`.

Reviewer: Replacing em-dash with a period.

```suggestion
- A **proof server**. Use your own (`docker run -p 6300:6300 midnightntwrk/proof-server:<tag>`) or the one the wallet reports via `getConfiguration()`.
```

  const t0 = Date.now();
  while (Date.now() - t0 < ms) {
    const p = pickWallet();
    if (p) return p;
    await new Promise((r) => setTimeout(r, 150));
  }
  return undefined;
};

const initial = await waitForWallet();
const api: ConnectedAPI = await initial!.connect('preview');   // prompts the user to approve
const config = await api.getConfiguration();                   // indexerUri, indexerWsUri, proverServerUri
const shielded = await api.getShieldedAddresses();             // shieldedCoinPublicKey, shieldedEncryptionPublicKey
```

## 4. Build the providers (wallet-delegated)

The key step: `balanceTx` serializes the unbound transaction and passes it to the wallet's `balanceUnsealedTransaction`. The wallet balances, proves, and uses its own synced state, then returns the finalized transaction. Your app never syncs the chain.

```ts
import { FetchZkConfigProvider } from '@midnight-ntwrk/midnight-js-fetch-zk-config-provider';
import { httpClientProofProvider } from '@midnight-ntwrk/midnight-js-http-client-proof-provider';
import { indexerPublicDataProvider } from '@midnight-ntwrk/midnight-js-indexer-public-data-provider';

> import { Binding, FinalizedTransaction, Proof, SignatureEnabled, Transaction } from '@midnight-ntwrk/ledger-v8';

Reviewer: `FinalizedTransaction` is a type alias in `@midnight-ntwrk/ledger-v8`, not a runtime class. Verified with `tsc --noEmit`. The import works because TypeScript erases it during compilation, but the inline `type` keyword signals intent and prevents accidental runtime references.

```suggestion
import { Binding, type FinalizedTransaction, Proof, SignatureEnabled, Transaction } from '@midnight-ntwrk/ledger-v8';
```

import { fromHex, toHex } from '@midnight-ntwrk/compact-runtime';
import type { UnboundTransaction } from '@midnight-ntwrk/midnight-js-types';

const zk = new FetchZkConfigProvider<string>(window.location.origin, fetch.bind(window));

const providers = {

>   privateStateProvider: inMemoryPrivateStateProvider(), // a simple in-memory implementation is fine for a demo

Reviewer: `inMemoryPrivateStateProvider()` is called here but never imported or defined anywhere in the tutorial. A reader following the steps will hit a `ReferenceError`. Consider either adding the import (and the package that exports it) or providing a minimal inline implementation so the code block is self-contained.

  zkConfigProvider: zk,
  proofProvider: httpClientProofProvider(config.proverServerUri!, zk),
  publicDataProvider: indexerPublicDataProvider(config.indexerUri, config.indexerWsUri),
  walletProvider: {
    getCoinPublicKey: () => shielded.shieldedCoinPublicKey,
    getEncryptionPublicKey: () => shielded.shieldedEncryptionPublicKey,
    balanceTx: async (tx: UnboundTransaction): Promise<FinalizedTransaction> => {
      const received = await api.balanceUnsealedTransaction(toHex(tx.serialize()));
      return Transaction.deserialize<SignatureEnabled, Proof, Binding>('signature', 'proof', 'binding', fromHex(received.tx));
    },
  },
  midnightProvider: {
    submitTx: async (tx: FinalizedTransaction) => {
      await api.submitTransaction(toHex(tx.serialize()));
      return tx.identifiers()[0];
    },
  },
};
```

## 5. Deploy a contract

```ts
import { CompiledContract } from '@midnight-ntwrk/compact-js';
import { deployContract } from '@midnight-ntwrk/midnight-js-contracts';
import * as MyContract from './managed/<name>/contract/index.js';

> const compiled = CompiledContract.make('myContract', (MyContract as any).Contract).pipe(

Reviewer: Six `as any` casts across sections 4-7 (`(MyContract as any).Contract`, `providers as any`, `(deployed as any).callTx`, `({ privateState }: any)`, `(MyContract as any).ledger`). These work but make the examples harder to adapt since a reader cannot see the expected types.

The zk-loan tutorial (`docs/tutorials/zk-loan/cli.mdx:460`) achieves the same pattern with a type parameter on `CompiledContract.make<ZKLoanCreditScorerContract>(...)`, which preserves type safety through the chain. Consider adding the contract type parameter where possible.

      crypto.getRandomValues(b);
      return [privateState, b];
    },
  }),
  CompiledContract.withCompiledFileAssets('myContract'),
);

const deployed = await deployContract(providers as any, {
  compiledContract: compiled,
  privateStateId: 'myPrivateState',
  initialPrivateState: {},
});

console.log('contract address:', deployed.deployTxData.public.contractAddress);
```

## 6. Call a circuit

Use `findDeployedContract` to attach to an existing contract, then call any impure circuit by name through `callTx`:

```ts
import { findDeployedContract } from '@midnight-ntwrk/midnight-js-contracts';

const deployed = await findDeployedContract(providers as any, {
  contractAddress,
  compiledContract: compiled,
  privateStateId: 'myPrivateState',
  initialPrivateState: {},
});

const txData = await (deployed as any).callTx.myCircuit();
console.log('tx hash:', txData.public.txHash);
```

## 7. Read contract state (no wallet required)

Reading on-chain state is public. Query the indexer and decode the result with the contract's generated `ledger` function:

```ts
const pdp = indexerPublicDataProvider(
  'https://indexer.preview.midnight.network/api/v3/graphql',

>   'wss://indexer.preview.midnight.network/api/v3/graphql/ws',

Reviewer: The current indexer API version is v4. The indexer 4.0.0 release notes state: "v3 is available as an alias of v4 to support migration windows for existing clients." All other tutorials use v4 (verified against `docs/tutorials/bboard/bboard-cli-implementation.mdx:192`).

```suggestion
  'https://indexer.preview.midnight.network/api/v4/graphql',
  'wss://indexer.preview.midnight.network/api/v4/graphql/ws',
```

);

const state = await pdp.queryContractState(contractAddress);
const led = (MyContract as any).ledger(state!.data);
// read your ledger fields, for example a Counter or a Map:
console.log('size:', led.items?.size?.());
```

## Troubleshooting

> - **Faucet returns "resubmit later" on a brand-new address.** The faucet rate-limit is keyed to the recipient address (about 24 hours), and a failed request still counts. Use a fresh address to retry. To confirm a drip landed without a wallet, query the indexer's `unshieldedTransactions(address:)` subscription, or `transactions(offset: { identifier })` if you have the transaction id — the top-level GraphQL API has no address-balance query.

Reviewer: Replacing em-dash with period. This sentence also runs quite long (56 words). Consider splitting after "retry" into a separate sentence.

```suggestion
- **Faucet returns "resubmit later" on a brand-new address.** The faucet rate-limit is keyed to the recipient address (about 24 hours), and a failed request still counts. Use a fresh address to retry. To confirm a drip landed without a wallet, query the indexer's `unshieldedTransactions(address:)` subscription, or `transactions(offset: { identifier })` if you have the transaction id. The top-level GraphQL API has no address-balance query.
```

- **`window.midnight` is undefined even though the extension is installed.** The extension's content-script/background messaging can stall (the console may show `orphaned data for stream` messages). Fully restart the browser to revive the extension's background worker, then reload. Also confirm the dApp tab and the extension are in the same browser profile.

> - **A helper library fails to resolve under bare Node** because it uses extensionless ESM imports. Bundle through Vite or esbuild rather than running with raw `node` — both resolve extensionless imports.

Reviewer: Replacing em-dash with period.

```suggestion
- **A helper library fails to resolve under bare Node** because it uses extensionless ESM imports. Bundle through Vite or esbuild rather than running with raw `node`. Both tools resolve extensionless imports.
```

- **Version drift.** The `compact-runtime` your contract was compiled against must match the SDK line you import (here 0.15 with ledger v8 and `midnight-js` 4.x). Mismatches between `compactc`, the language version, and the runtime are the most common source of errors.

## Summary

> You connected an injected wallet, built providers that delegate balancing and signing to it, then deployed a contract, called a circuit, and read state — all from the browser, without running a headless wallet sync. The same pattern works for any Compact contract: swap in your compiled bindings and ZK assets, and call your own circuits by name.

Reviewer: Replacing em-dash with comma. Also, the style guide recommends tutorials end with links to related content. Consider adding a brief "Next steps" section after the summary, pointing to the leaderboard tutorial (which extends this pattern into a full multi-page DApp) or the SDK reference docs.

```suggestion
You connected an injected wallet, built providers that delegate balancing and signing to it, then deployed a contract, called a circuit, and read state, all from the browser, without running a headless wallet sync. The same pattern works for any Compact contract: swap in your compiled bindings and ZK assets, and call your own circuits by name.
```
